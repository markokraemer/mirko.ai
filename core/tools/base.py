from dataclasses import dataclass
from abc import ABC, abstractmethod
import json


@dataclass
class ToolResult:
    success: bool
    exit_code: int
    output: str

class Tool(ABC):
    @abstractmethod
    def schema(self) -> list[dict]:
        pass

    def success_response(self, data: dict | str) -> ToolResult:
        if isinstance(data, str):
            text = data
        else:
            text = json.dumps(data, indent=2)
        return ToolResult(success=True, exit_code=0, output=text)

    def fail_response(self, msg: str, code: int = 1) -> ToolResult:
        return ToolResult(success=False, exit_code=code, output=msg)



