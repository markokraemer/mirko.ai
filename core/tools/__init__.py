from .terminal_tool import TerminalTool
from .retrieval_tool import RetrievalTool
from .base import ToolResult