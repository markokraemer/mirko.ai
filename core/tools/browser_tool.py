from selenium import webdriver
from selenium.webdriver.common.by import By
from core.memory.working_memory import WorkingMemory
from .base import Tool, ToolResult
import time

class BrowserTool(Tool):
    def __init__(self):
        """
        Initializes the BrowserTool with specific Chrome webdriver options.
        """
        # Initialize the webdriver options as in the provided context
        self.options = webdriver.ChromeOptions()
        self.options.add_argument('headless')
        self.options.add_argument('window-size=1200x600')
        self.options.add_argument('--no-sandbox')
        self.options.add_argument('--disable-dev-shm-usage')
        self.options.add_argument('--disable-gpu')
        self.options.add_argument('--remote-debugging-port=9222')
        self.options.add_experimental_option('excludeSwitches', ['enable-logging'])

    def check_browser_tab_for_error(self, full_page_url):
        """
        Checks for errors in a browser tab by loading a URL and inspecting browser logs and page content.

        Args:
            full_page_url (str): The URL of the page to check for errors.

        Returns:
            ToolResult: A ToolResult object indicating whether errors were found, along with the error details.
        """
        driver = webdriver.Chrome(options=self.options)
        driver.get(full_page_url)
        time.sleep(6)  # Wait for the page to load
        logs = driver.get_log('browser')
        errors = [log for log in logs if log['level'] == 'SEVERE']  # Filter out 'SEVERE' level logs as errors

        # Attempt to read the page content for visible error messages
        try:
            page_content = driver.find_element(By.TAG_NAME, 'body').text
            if "Error" in page_content:
                errors.append({'level': 'SEVERE', 'message': page_content, 'source': 'page-content'})
        except Exception as e:
            errors.append({'level': 'SEVERE', 'message': f"Failed to read page content: {str(e)}", 'source': 'page-content'})

        driver.quit()  # Ensure the driver is quit after operation
        if errors:
            return ToolResult(success=True, output=errors, exit_code=0)
        else:
            return ToolResult(success=False, output="No errors found.", exit_code=1)

    @staticmethod
    def schema() -> list[dict]:
        """
        Returns the OpenAPI JSON schema for function calls in the Browser Tool.
        """
        return [
            {
                "type": "function",
                "function": {
                    "name": BrowserTool.check_browser_tab_for_error.__name__,
                    "description": BrowserTool.check_browser_tab_for_error.__doc__,
                    "parameters": {
                        "type": "object",
                        "properties": {
                            "full_page_url": {
                                "type": "string",
                                "description": "The full URL of the page to check for errors.",
                            }
                        },
                        "required": ["full_page_url"],
                    },
                },
            },
        ]

# Example usage
if __name__ == "__main__":
    browser_tool = BrowserTool()
    errors = browser_tool.check_browser_tab_for_error("https://www.youtube.com/watch?v=LLxWakCWUAI")
    if errors.success:
        print("Errors found:")
        for error in errors.output:
            print(error)
    else:
        print("No errors found.")

