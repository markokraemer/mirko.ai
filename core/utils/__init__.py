from .workspace_utils import (
    initialise_workspace,
    get_docker_container_id,
    get_container_merged_dir,
) 

from .file_utils import (
    find_files,
)

from .llm import (
    make_llm_api_call,
    execute_tool_calls
)