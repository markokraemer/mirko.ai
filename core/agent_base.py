from typing import List, Dict
import os
from openai import OpenAI
import time
import json
import inspect
import asyncio
import logging

from core.tools import TerminalTool, FilesTool, ToolResult
from core.utils.llm import make_llm_api_call
from core.utils.debug_logging import initialize_logging

os.getenv("OPENAI_API_KEY")
client = OpenAI()
initialize_logging()



# =========================
# Base Assistant Class
# =========================

class BaseAssistant:
    def __init__(self, name: str, instructions: str, tools: List[Dict] = []):
        self.name = name
        self.instructions = instructions
        self.tools = tools
        self.assistant_id = self.create_assistant(name, instructions, tools)

    @staticmethod
    def create_assistant(name, instructions, tools=[], model="gpt-4-turbo-preview"):
        
        assistant = client.beta.assistants.create(
            name=name,
            instructions=instructions,
            tools=tools,
            model=model
        )
        return assistant.id

    @staticmethod
    def start_new_thread():
        empty_thread = client.beta.threads.create()
        return empty_thread.id

    @staticmethod
    def add_message(thread_id, content, role="user"):
        thread_message = client.beta.threads.messages.create(
            thread_id=thread_id,
            role=role,
            content=content,
        )
        return thread_message

    @staticmethod
    def run_thread(thread_id, assistant_id):
        run = client.beta.threads.runs.create(
            thread_id=thread_id,
            assistant_id=assistant_id,
        )
        return run.id

    @staticmethod
    def get_run(thread_id, run_id):
        run = client.beta.threads.runs.retrieve(
        thread_id=thread_id,
        run_id=run_id
        )
        return run

    @staticmethod
    def get_messages_in_thread(thread_id, stringified=False):
        messages = client.beta.threads.messages.list(thread_id)
        # Sort messages by 'created_at' to ensure they are in the correct order
        sorted_messages = sorted(messages.data, key=lambda x: x.created_at)
        if stringified:
            string_messages = []
            for message in sorted_messages:
                role = message.role.upper()
                # Assuming the first item in content list is the main text
                content = message.content[0].text.value
                string_messages.append(f"{role}: {content}")
            return "\n\n".join(string_messages)
        return sorted_messages

    async def check_run_status_and_execute_action(self, thread_id: str, run_id: str):
        while True:
            run = self.get_run(thread_id, run_id)
            if run.status == "requires_action":
                await self.execute_run_action(run_id, thread_id)
                await asyncio.sleep(3)
            elif run.status in ["in_progress", "queued", "cancelling"]:
                await asyncio.sleep(3)                
            elif run.status in ["completed", "cancelled", "failed", "expired"]:
                break

    async def execute_run_action(self, run_id, thread_id):
        run = BaseAssistant.get_run(thread_id, run_id)
        required_action = run.required_action if run.status == "requires_action" else None
        logging.info(f"Debug: Required action for run_id {run_id} is {required_action}")

        if required_action and required_action.type == "submit_tool_outputs":
            tool_calls = required_action.submit_tool_outputs.tool_calls
            tool_outputs = []
            logging.info(f"Debug: Processing {len(tool_calls)} tool calls for submission.")

            # Instantiate TerminalTool and FilesTool without hardcoded arguments
            terminal_tool = TerminalTool()
            files_tool = FilesTool()

            for tool_call in tool_calls:
                function_name = tool_call.function.name
                arguments = json.loads(tool_call.function.arguments)
                logging.info(f"Debug: Attempting to execute function {function_name} with arguments {arguments}")

                # Dynamically call the function with the provided arguments
                try:
                    # Attempt to find the function in terminal_tool or files_tool instances
                    function = None
                    for tool_instance in [terminal_tool, files_tool]:
                        if hasattr(tool_instance, function_name):
                            function = getattr(tool_instance, function_name)
                            logging.info(f"Debug: Found function {function_name} in {type(tool_instance).__name__}")
                            break

                    if function:
                        # Execute the function asynchronously if it's a coroutine
                        if inspect.iscoroutinefunction(function):
                            output = await function(**arguments)
                        else:
                            output = function(**arguments)
                        logging.info(f"Debug: Function {function_name} executed successfully with output: {output}")
                        # Extract the output if it's an instance of ToolResult
                        if isinstance(output, ToolResult):
                            output = output.output
                    else:
                        output = "Function not found"
                        logging.info(f"Debug: Function {function_name} not found in TerminalTool or FilesTool instances")
                except Exception as e:
                    output = f"An exception has occurred: {e}"
                    logging.info(f"Debug: Exception occurred while executing function {function_name}: {e}")

                tool_outputs.append({
                    "tool_call_id": tool_call.id,
                    "output": str(output)  # Convert output to string to avoid type error
                })

            # Submit the tool outputs
            try:
                run = client.beta.threads.runs.submit_tool_outputs(
                    thread_id=thread_id,
                    run_id=run_id,
                    tool_outputs=tool_outputs
                )
                logging.info(f"Debug: Successfully submitted tool outputs for run_id {run_id}")
            except Exception as e:
                logging.info(f"Failed to submit tool outputs for run_id {run_id}: {e}")

    def generate_next_action(self, thread_id):
        messages_in_thread = self.get_messages_in_thread(thread_id, stringified=True)
        messages = [
            {
                "role": "system",
                "content": """     
    You are the internal monologue of Mirko.ai the Expert AI Software Engineer. 

    Context: Mirko.ai is an advanced AI software engineer, capable of independent problem-solving and decision-making within the realm of software development. Mirko.ai has been assigned with a user objective and is working iteratively to accomplish it.

    As the internal monologue of Mirko.ai, you are reflecting on your current situation and planning your next steps towards achieving the user objective.

    Output a JSON with: 
    - Your observations, thoughts in a tree-of-thoughts, upcoming actions list, next action to conduct.

 """
            },
            {
                "role": "user",
                "content": f""" Event History: {messages_in_thread} 
                Working Memory: 
"""
            },
        ]
        response = make_llm_api_call(messages, model_name="gpt-4-turbo-preview")
        new_message_contents = response.choices[0].message['content']
        self.add_message(thread_id, new_message_contents, role="user")


    def generate_playground_access(self, thread_id):
        playground_url = f'https://platform.openai.com/playground?assistant={self.assistant_id}&mode=assistant&thread={thread_id}'
        logging.info(f'Playground Access URL: {playground_url}')




async def start_session_run(user_request):

    tools = []
    tools.extend(TerminalTool.schema())
    tools.extend(FilesTool.schema())

    instructions = """
    You are Mirko, an expert AI software engineer, a brilliant and meticulous software engineer working iteratively towards implementing the received task from the user. You are an autonomous cognitive digital entity using your tools to interact with your personal Linux workspace.
    
    
    - Generate_task_list
    - Process_task
    - Send_web_browsing_instructions_to_browsing_agent(he reports back)
    - Send_terminal_commands
    - WM: Full Latest Directory Contents in Session, Open Tabs & past interactions, Open Terminal Sessions and most recent log, 
    
    Retrieval:
    - Expand_File_tree
    - Open_File
    - Open_Files_in_Editor
    - Read_directory_contents
    - Read_file_contents

    - Code_snippet_retrieval(query)

    Files:
    - Rename_file(current_file_path, new_file_path)
    - Delete_file(file_path)
    - Move_file(current_file_path, new_file_path)
    - Edit_file_contents(instructions)

    Terminal:
    - New_terminal_session()
    - Close_terminal_session(session_id)
    - Send_terminal_command(session_id, command)
    - get_session_log(session_id, timestamp/other-indicator)

    Browsing:
    - Send_browsing_instructions()


    dissect, think step by step... etc.. – have it as function calls too

    """    

    assistant = BaseAssistant("Mirko.ai", instructions, tools)

    # Initial User Request
    thread_id = assistant.start_new_thread()
    assistant.generate_playground_access(thread_id)
    assistant.add_message(thread_id, user_request, "user")
    # Recursive self-debate Loop
    while True:
        run_id = assistant.run_thread(thread_id, assistant.assistant_id)
        await assistant.check_run_status_and_execute_action(thread_id, run_id)
        assistant.generate_next_action(thread_id)

        # Sample usage of get_messages_in_thread with stringified=True
        stringified_messages = assistant.get_messages_in_thread(thread_id, stringified=True)
        logging.info(f"Stringified message thread {thread_id} for {assistant.name}: {stringified_messages}")


if __name__ == "__main__":

    user_request = "Build a simple Landing Page for my construction company. "        
    asyncio.run(start_session_run(user_request))


