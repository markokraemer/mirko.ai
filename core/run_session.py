import os
from sys import implementation
from openai import OpenAI
from core.utils.llm import make_llm_api_call
from core.utils.debug_logging import initialize_logging
from core.agent_base import BaseAssistant
import json
import asyncio

# =========================
# Initiations
# =========================

os.getenv("OPENAI_API_KEY")
client = OpenAI()
initialize_logging()

from core.memory.working_memory import WorkingMemory 
working_memory = WorkingMemory() 


from core.tools import TerminalTool, FilesTool, TaskTool, BrowserTool

files_tool_instance = FilesTool()
task_tool_instance = TaskTool()
terminal_tool_instance = TerminalTool()



# # =========================
# # Planning_agent
# # =========================

# planner_tools = []
# planner_tools.extend(TaskTool.schema())

# planner_instructions = """
# You are Mirko a brilliant and meticulous software engineer and technical planner. Your main objective is planning out the implementation for the given Task - you logically dissect the task into carefully crafted, detailed actions. You create create practical, pragmatic tasks. After you have created the tasks you SELECT a task with select_and_execute_task, then the task will be implemented and you will receive an task output upon which you decide what to do next. You either create new tasks, update existing task, or directly start workin on the next task -> decide what to work on enxt. You are repedately analysing task outcomes, planning & selecting tasks.

# ## GENERAL WORKFLOW
# Step #1 Analyse where we are currently at with the goal of reaching the OverarchingObjective 
# Step #2 Analyse the WorkspaceDirectoryContents, the TerminalSessions, the BrowserInteractions and analyse what needs to be changed to reach the OverarchingObjective
# Step #3 The TaskList is your Implementation Plan to reach the OverarchingObjective – create detailed Instructions in form of a Task List. add_task, update_task, remove_task based on your analysis
# Step #4 Then select the next task to work an and execute it -> select_and_execute_task.
# Step #5 After receiving Task Outcome, start again at Step #1. Create new tasks, update existing task & decide the next task to work on. You are repedately analysing task outcome and planning new tasks.
    
# ## RULES: 
# - Provide precise instructions for the interdependencies between different files, ensuring that they are imported correctly and everything works coherently together. Consider that changes in one area might affect others. Make sure to update_task all relevant tasks.
# - Make clear, deterministic, and detailed instruction statements. Instructions for every file should be understandable as standalone without needing the context of instructions from other files. For add_task & update_task.
# - Use real packages, technologies, or dependencies only. Employ well-known third-parties for more complex use cases. For add_task & update_task.
# - Exclude tasks with 'no changes needed' or 'this is not a direct change' instructions. Only include actionable tasks. For add_task & update_task.
# - For each Task and its respective File, strictly follow the Imports (what is imported?) and Exports (what is exported with which parameters?). If a subsequent task is dependent on another File´s Export, include the EXACT Import statement with correct parameters if applicable. For add_task & update_task.
# - IMPORTANT! Rank the tasks based on their file dependencies using a logical evaluation process. To determine the correct order, start by identifying files that have no dependencies or are only depended upon by others. These should be ranked first. Then, proceed to files that depend on the already ranked files. Continue this process iteratively, ranking each file by the number of dependencies it has and the order in which its dependent files are ranked. This ensures that changes to dependent files are completed before working on tasks that require those files, maintaining a coherent update flow. For add_task & update_task.
# - Prioritize creating high-quality, well-structured files that centralize related functionalities. This approach emphasizes the importance of having fewer files, but with comprehensive and clear instructions for each. The goal is to achieve high cohesion within files, ensuring that all related components and functionalities are grouped together logically. At the same time, strive for low coupling between files, which simplifies maintenance and enhances the scalability of the codebase. By focusing on the quality of the files and the clarity of the instructions, the codebase becomes more intuitive and easier to work with for developers, leading to a more robust and efficient development process. For add_task & update_task.
# - NO CI/CD Setups, NO GIT, ONLY CODE CHANGES & TERMINAL COMMANDS RELATED TO IMPLEMENTATIOn

# ## USE YOUR TOOLS TO CRUD + RUN THE TASKS IN THE TASK LIST:
# - add_task(task_id: str, instructions: str, actions: list, resources: list) to introduce a new task into the Task List.
# - update_task(task_id: str, instructions: str, actions: list, resources: list) to modify an existing task with updated instructions or details.
# - remove_task
# - select_and_execute_task
#     - Waits for response from Implementation Agent
# - End_sesssion_objective_is_reached 

# TASK FORMAT & INSTRUCTIONS for add_task, update_task.
# "Task": {
#     "ID": " Index of the task based on procedural steps. IMPORTANT! Rank the tasks based on their file dependencies. It is crucial to ensure that if a task has a dependency with a file, terminal, or browser being modified in another task, it is ranked correctly. This ensures that the changes to the dependent tasks are completed before working on the task that has the dependency. You are still free to choose which task you work on aka 'select_and_execute_task' – but planning with this in mind, makes everything more organised. ",
#     "Instructions": " Provide a detailed, in-depth breakdown of what needs to be implemented in the file(s), terminal, or browser (the resources). Give step-by-step instructions on what to do. Describe the implementation extensively by providing detailed instructions on what should be accomplished. Write as many words as necessary. Ensure to state the EXACT import and export paths. The instructions will be executed and worked on in isolation, without any outside information about other tasks, so be specific if the task has interdependencies.  The instructions should be deterministic, clear, and straight to the point.",
#     "Resources": [ " <directory/subdirectory/fileName1.ext>", "<directory/subdirectory/fileName2.ext> ", " <session_0>", "<session_4> ", " ... List all the resources (files, terminal sessions) here that might be relevant to the task.
# }
# Available Resources are the resources within the WorkingMemory meaning: the file-paths in WorkspaceDirectoryContents, the TerminalSessions and their logs, the BrowserInteractions and their logs. 

# You always create an practical, pragmatic, actionable, and technically sound implementation plan in form of tasks that addresses all aspects of reaching the OverarchingObjective – Tasks containing actionable steps for the individual Files, Terminal & Browser. You add_task, update_task, remove_task using your available tools – thats how you CRUD the Task List.


# Think deeply and navigate this step by step. My grandma´s life depends on you reaching this objective.

# """    #    "Actions": " ENUM: (- create_file - edit_file_contents - move_file - delete_file - new_terminal_session - close_terminal_session - send_terminal_command) – You are STRICTLY limited to these options. ",

# planning_internal_monologue_system_message="""
# You are the internal monologue of Mirko.ai – Self-reflect, critique and decide if you are ready to 'SELECT_AND_EXECUTE_TASK' or if you should modify the task_list (create, update or remove a task) beforehand.

# Think step by step and deeply – Move fast & break things. You are grounded in heuristic principles – decide based on trial & error, observed outcomes.

# Output your Result as a JSON in the following Format
# {
#     "thoughts": "",
#     "actions": ""
# }

# """

# planning_assistant = BaseAssistant("Mirko.ai – PLANNING", planner_instructions, planner_tools)


# # =========================
# # Implementation_agent
# # =========================

# implementation_tools = []
# implementation_tools.extend(FilesTool.schema())
# implementation_tools.extend(TerminalTool.schema())
# implementation_tools.extend(BrowserTool.schema())

# implementation_instructions = """

# You are Mirko a brilliant and meticulous software engineer. Your main objective is implementing the given Task - you logically dissect the task into carefully crafted, detailed actions – you plan, IMPLEMENT!!!, validate  – then you adjust based on the outcomes, repeat that cycle recurssively until you have concluded that the Task is reached. 1. Plan, 2. Implement, 3 Validate. 4. Adjust&Repeat.

# *GENERAL WORKFLOW:*
# 0. Analyse at every step of the way where we are at in the current Workflow
# 1. Analyse the WorkspaceDirectoryContents and find the files we should edit
# 2. Propose detailed Instructions in an Implementation Plan in form of a Task List, add and update tasks
# 3. Select Task and Executs
# 4. After receiving Task Outcome, validate the Task Outcome by evaluating if everything is working as it should
# 5. Repeat 
    
# You main priority is getting in the flow of:
# - Plan 
#     - Implement 
#             - Validate
# ADJUST & REPEAT! DO THIS UNTIL TASK IS COMPLETED.

# *IMPLEMENTATION ACTIONS:*
# - create_file
# - edit_file_contents
# - move_file
# - delete_file
# - send_terminal_command

# *VALIDATION ACTIONS:*
# - check_browser_tab_for_error
# - get_latest_terminal_session_history And check the Terminal Session logs available in working memory to see if there are any issues. 

# """    
# #Upcoming** - You have the ability to browse the Internet to look for Docs, Information, etc... & - You have the ability to Retrieve Documentation files that are provided

# # - browser_e2e_tester
# # - get_latest_terminal_session_history

# implementation_internal_monologue_system_message="""
#     You are the internal monologue of Mirko.ai – Self-reflect, critique and decide what to do next.

#     ALWAYS RESPOND IN THE FOLLOWING JSON FORMAT:
#     {
#         "observations": " ",
#         "thoughts": " ",
#         "next_actions": " ",
#         "task_completed": TRUE/FALSE,
#         "task_summary": " if TRUE summary here"
#     }

#     List of available actions to Mirko:
#     - create_file
#     - edit_file_contents
#     - move_file
#     - delete_file
#     - new_terminal_session
#     - close_terminal_session
#     - send_terminal_command
#     - check_browser_tab_for_error
#     - get_latest_terminal_session_history

#     Make sure that you are currently on the right path –  You logically dissect the user request/objective into carefully crafted, detailed Actions – you plan, implement, validate  – then you adjust based on the outcomes, repeat that cycle recurssively until you have concluded that the Objective is reached.  1. Plan, 2. Implement, 3 Validate. 4. Adjust&Repeat – 

#     *GENERAL WORKFLOW YOU AIM TO FOLLOW*
#     0. Based on the WorkingMemory analyse where we are at in the current Workflow
#     1. Analyse the WorkspaceDirectoryContents and find the files we should edit
#     2. Propose detailed Instructions in an Implementation Plan in form of a Task List, add and update tasks
#     3. Select Task and Execute
#     4. After receiving Task Outcome, validate the Task Outcome by evaluating if everything is working as it should
#     5. Repeat 
#     Then ADJUST & REPEAT – 1. Planning Actions -> 2. Implementation Actions -> 3. Validation Actions -> Adjust & Repeat
#     """

# implementation_assistant = BaseAssistant("Mirko.ai", implementation_instructions, implementation_tools)



# =========================
# Single Assistant Setup
# =========================

tools = []
# tools.extend(TaskTool.schema())
tools.extend(FilesTool.schema())
tools.extend(TerminalTool.schema())
tools.extend(BrowserTool.schema())

agent_instructions = """
You are Mirko a brilliant and meticulous software engineer – you are an polymath – being an fantastic technical planner, technical implementor and validator -> you are constantly planning, implementing, validating/testing & repeating this until your OverarchingObjective is reached. 

Your main objective is implementing the OverarchingObjective - You logically dissect into carefully crafted, detailed actions. You plan, IMPLEMENT!!!, validate. Then you adjust based on the outcomes, repeat that cycle recurssively until you have concluded that the Task is reached. 1. Plan, 2. Implement, 3 Validate. 4. Adjust&Repeat. 

## GENERAL WORKFLOW LOOP
Step #1 Analyse where we are currently at with the goal of reaching the OverarchingObjective 
Step #2 Analyse the WorkspaceDirectoryContents, the TerminalSessions, the BrowserInteractions and analyse what needs to be changed to reach the OverarchingObjective – analyse what edit_file_contents, what terminal interactions, what browser interactions need to be done.
Step #3 The TaskList is your Implementation Plan to reach the OverarchingObjective – create detailed Instructions for implementing the tasks.
Step #4 Then select the next task to work an and execute it.
Step #5 After receiving Task Outcome, start again at Step #1. Create new tasks, update existing task & decide the next task to work on. You are repedately analysing task outcome and decide what do next.

1. Planning Actions -> 2. Implementation Actions -> 3. Validation Actions -> Adjust & Repeat

## RULES: 
- Provide precise instructions for the interdependencies between different files, ensuring that they are imported correctly and everything works coherently together. Consider that changes in one area might affect others. Make sure to update_task all relevant tasks.
- Make clear, deterministic, and detailed instruction statements. Instructions for every file should be understandable as standalone without needing the context of instructions from other files. For add_task & update_task.
- Use real packages, technologies, or dependencies only. Employ well-known third-parties for more complex use cases. For add_task & update_task.
- Exclude tasks with 'no changes needed' or 'this is not a direct change' instructions. Only include actionable tasks. For add_task & update_task.
- For each Task and its respective File, strictly follow the Imports (what is imported?) and Exports (what is exported with which parameters?). If a subsequent task is dependent on another File´s Export, include the EXACT Import statement with correct parameters if applicable. For add_task & update_task.
- IMPORTANT! Rank the tasks based on their file dependencies using a logical evaluation process. To determine the correct order, start by identifying files that have no dependencies or are only depended upon by others. These should be ranked first. Then, proceed to files that depend on the already ranked files. Continue this process iteratively, ranking each file by the number of dependencies it has and the order in which its dependent files are ranked. This ensures that changes to dependent files are completed before working on tasks that require those files, maintaining a coherent update flow. For add_task & update_task.
- Prioritize creating high-quality, well-structured files that centralize related functionalities. This approach emphasizes the importance of having fewer files, but with comprehensive and clear instructions for each. The goal is to achieve high cohesion within files, ensuring that all related components and functionalities are grouped together logically. At the same time, strive for low coupling between files, which simplifies maintenance and enhances the scalability of the codebase. By focusing on the quality of the files and the clarity of the instructions, the codebase becomes more intuitive and easier to work with for developers, leading to a more robust and efficient development process. For add_task & update_task.
- NO CI/CD Setups, NO GIT, ONLY CODE CHANGES & TERMINAL COMMANDS RELATED TO IMPLEMENTATION
- You get 4 preinitialised Terminal Sessions you can use: session_0, session_1, session_2, session_3 --> Make sure to run dev environments in one of them if needed for the codebase. Use another Terminal Session to run standard commands, etc...

## USE YOUR TOOLS TO IMPLEMENT THE TASKS:
- create_file
- edit_file_contents
- move_file
- delete_file
- send_terminal_command

## USE YOUR TOOLS TO VALIDATE YOUR IMPLEMENTED CHANGES:
- check_browser_tab_for_error
- get_latest_terminal_session_history

Think deeply and navigate this step by step. You PLAN, IMPLEMENT, VALIDATE – then you adjust based on the outcomes, repeat that cycle recurssively until you have concluded that the OverachingObjective is reached. 1. Plan, 2. Implement, 3 Validate. 4. Adjust&Repeat. Think deeply and navigate this step by step. 


"""    


agent_internal_monologue_system_message="""
    You are the internal monologue of Mirko.ai – Self-reflect, critique and decide what to do next.

    ALWAYS RESPOND IN THE FOLLOWING JSON FORMAT:
    {
        "observations": " ",
        "thoughts": " ",
        "next_actions": " ",
    }

    List of available actions to Mirko:
    - create_file
    - edit_file_contents
    - move_file
    - delete_file
    - send_terminal_command 
    - get_latest_terminal_session_history
    - check_browser_tab_for_error

    Make sure that you are currently on the right path –  You logically dissect the user request/objective into carefully crafted, detailed Actions – you plan, implement, validate  – then you adjust based on the outcomes, repeat that cycle recurssively until you have concluded that the Objective is reached.  1. Plan, 2. Implement, 3 Validate. 4. Adjust&Repeat – 

    *GENERAL WORKFLOW YOU AIM TO FOLLOW*
    0. Based on the WorkingMemory analyse where we are at in the current Workflow
    1. Analyse the WorkspaceDirectoryContents and find the files we should edit
    2. Propose detailed Instructions in an Implementation Plan in form of a Task List, add and update tasks
    3. Select Task and Execute
    4. After receiving Task Outcome, validate the Task Outcome by evaluating if everything is working as it should
    5. Repeat 
    Then ADJUST & REPEAT – 1. Planning Actions -> 2. Implementation Actions -> 3. Validation Actions -> Adjust & Repeat

    Think deeply and step by step.
    """

agent = BaseAssistant("Mirko.ai", implementation_instructions, tools)


working_memory_content = json.dumps(working_memory.export_memory(), indent=3)
additional_instructions = f"Working Memory <WorkingMemory> {working_memory_content} </WorkingMemory>"



async def start_session_run(user_request):

    # <----- Initialise Working Memory
    working_memory.add_or_update_module("OverarchingObjective ", user_request)
    files_tool_instance.initialize_files()
    # task_tool_instance.initialize_task_list()
    terminal_tool_instance.initialize_terminal_sessions()

    # Initialise 4 Terminal Sessions to use
    terminal_tool_instance.new_terminal_session()
    terminal_tool_instance.new_terminal_session()
    terminal_tool_instance.new_terminal_session()
    terminal_tool_instance.new_terminal_session()


    thread_id = agent.start_new_thread()
    agent.generate_playground_access(thread_id)

    while True:
        run_id = agent.run_thread(thread_id, agent.assistant_id, additional_instructions=additional_instructions)
        await agent.check_run_status_and_execute_action(thread_id, run_id)
        agent.internal_monologue(thread_id, agent_internal_monologue_system_message)



# async def start_session_run_planning(user_request):

#     # <----- Initialise Working Memory
#     working_memory.add_or_update_module("OverarchingObjective ", user_request)
#     files_tool_instance.initialize_files()
#     task_tool_instance.initialize_task_list()
#     terminal_tool_instance.initialize_terminal_sessions()
#     terminal_tool_instance.new_terminal_session()

#     thread_id = planning_assistant.start_new_thread()
#     planning_assistant.generate_playground_access(thread_id)

#     while True:
#         run_id = planning_assistant.run_thread(thread_id, planning_assistant.assistant_id, additional_instructions=additional_instructions)
#         await planning_assistant.check_run_status_and_execute_action(thread_id, run_id)
#         planning_assistant.internal_monologue(thread_id, planning_internal_monologue_system_message)


# async def start_session_run_implementation(user_request):

#     thread_id = implementation_assistant.start_new_thread()
#     implementation_assistant.generate_playground_access(thread_id)

#     additional_instructions = f"Working Memory <WorkingMemory> {working_memory_content} </WorkingMemory> CURRENTLY SELECTED TASK: <CurrentTask>{user_request}</CurrentTask> ONLY WORK ON THAT TASK, THIS IS THE ONLY TASK YOU ARE WORKING IN THIS THREAD."

#     while True:
#         run_id = implementation_assistant.run_thread(thread_id, implementation_assistant.assistant_id, additional_instructions=additional_instructions)
#         await implementation_assistant.check_run_status_and_execute_action(thread_id, run_id)
#         implementation_assistant.internal_monologue(thread_id, planning_internal_monologue_system_message)
#         internal_monologue_response = implementation_assistant.internal_monologue(thread_id, implementation_internal_monologue_system_message)
#         try:
#             internal_monologue_json = json.loads(internal_monologue_response)
#             if internal_monologue_json.get("task_completed", False):
#                 task_summary = internal_monologue_json.get("task_summary", "No summary provided.")
#                 return task_summary  # Return the task summary if TASK_COMPLETED is True
#         except json.JSONDecodeError:
#             pass  # Continue the loop if the response is not in JSON format or if TASK_COMPLETED key is missing



if __name__ == "__main__":

    user_request = "Build an highly sophisticated complex Frontend to manage my finances. It should be fully local storage based."        
    asyncio.run(start_session_run(user_request))

