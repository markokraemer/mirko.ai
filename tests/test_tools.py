import os
import json
from core.tools import TerminalTool, RetrievalTool, get_container_merged_dir
from core.utils.file_utils import find_files
from approvaltests.approvals import verify

def assert_tool_fields(schema):
    """
    Check some basic fields on an OpenAI function call tool definition.
    If we wanted to be rigorious we could use the jsonschema, but this is a start.
    """
    for tool in schema: 
        assert tool['type'] == 'function'
        assert tool['function']
        assert tool['function']['name']
        assert tool['function']['description']
        assert tool['function']['parameters']

def test_terminal_schema():
    subject = TerminalTool.schema()
    assert_tool_fields(subject)
    verify(subject)

def test_retrieval_schema():
    subject = RetrievalTool.schema()
    assert_tool_fields(subject)
    verify(subject)


def test_find_files(fs):
    # The fs fixture is pyfakefs
    fs.create_file("/var/fakedata/a/xx1.txt")
    fs.create_file("/var/fakedata/a/.DS_Store") # Excluded
    
    found = find_files("/var/fakedata", 1)
    assert found == ["a/xx1.txt"]

def test_find_files_depth_limit(fs):
    fs.create_file("/var/fakedata/a/b/xx1.txt")
    found = find_files("/var/fakedata", 1)
    assert found == []

def test_find_files_exclude_ext(fs):
    fs.create_file("/var/fakedata/xx1.jpg")
    found = find_files("/var/fakedata", 1)
    assert found == []

def test_find_files_exclude_dirname(fs):
    fs.create_file("/var/fakedata/node_modules/xx1.jpg")
    found = find_files("/var/fakedata", 3)
    assert found == []

def test_retrieval_Tool_read_file_contents(fs):
    fs.create_file("/var/fakedata/a.txt", contents="..contents..")
    subject = RetrievalTool("/var/fakedata")
    found = subject.read_file_contents("a.txt")
    assert found.ok, found.output
    assert found.output == '..contents..'

if os.getenv("INTEGRATION"):
    def test_send_ls():
        print(TerminalTool("workspace-dev-env-1").send_command("ls"))

    def test_list_tree():
        base_path = get_container_merged_dir("workspace-dev-env-1")
        subject = RetrievalTool(base_path)
        print(subject.get_file_tree(".", 2))

        