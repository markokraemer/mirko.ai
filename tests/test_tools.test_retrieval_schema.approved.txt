[{'type': 'function', 'function': {'name': 'get_file_tree', 'description': '\n        List path recursively with limited depth. Path is relative to current dir.\n        ', 'parameters': {'type': 'object', 'properties': {'path': {'type': 'string', 'description': 'Directory to list relative to current'}, 'depth': {'type': 'number', 'description': 'Depth limit, default 1'}}, 'required': ['path']}}}, {'type': 'function', 'function': {'name': 'read_file_contents', 'description': '\n        Show the contents of the file at path relative to current dir\n        ', 'parameters': {'type': 'object', 'properties': {'path': {'type': 'string', 'description': 'Path of file relative to current dir'}}, 'required': ['path']}}}]
